mergin/test/test_client.py

```diff
@@ -2003,3 +2011,53 @@ def test_clean_diff_files(mc):
     diff_files = glob.glob("*-diff-*", root_dir=os.path.split(mp.fpath_meta("inserted_1_A.gpkg"))[0])
 
     assert diff_files == []
+
+
+def test_download_failure(mc):
+    test_project = "test_download_failure"
+    project = API_USER + "/" + test_project
+    project_dir = os.path.join(TMP_DIR, test_project)
+    download_dir = os.path.join(TMP_DIR, "download", test_project)
+
+    cleanup(mc, project, [project_dir, download_dir])
+    # prepare local project
+    shutil.copytree(TEST_DATA_DIR, project_dir)
+
+    # create remote project
+    mc.create_project_and_push(test_project, directory=project_dir)
+
+    # download project async
+    with pytest.raises(IsADirectoryError):
+        job = download_project_async(mc, project, download_dir)
+        os.makedirs(os.path.join(download_dir, "base.gpkg.0"))
+        download_project_wait(job)
+        download_project_finalize(job)
+
+    assert job.failure_log_file is not None
+    with open(job.failure_log_file, "r", encoding="utf-8") as f:
+        content = f.read()
+        assert "Traceback" in content
+
+    # active waiting
+    remove_folders([download_dir])
+    job = download_project_async(mc, project, download_dir)
+    os.makedirs(os.path.join(download_dir, "base.gpkg.0"))
+    is_running = True
+    while is_running:
+        time.sleep(1)
+        try:
+            is_running = download_project_is_running(job)
+        except Exception:
+            download_project_cancel(job)
+            break
```

Comment:
I would suggest to simplify the whole block so that we wait for the exception indefinitely, and we do not expect that the download could succeed - to something like this?
```python
with pytest.raises(IsADirectoryError):
  while True:
    assert download_project_is_running(job)

download_project_cancel(job)  # to clean up things
```